### ISSUE TITLE
Command line mode runs as a separate process, detached from console

### ISSUE THREAD

**Issue by himselfv** [Bionus/imgbrd-grabber#479]
Grabber 4.2.2, fresh installation.

How to reproduce:

1. Open console.
2. Grabber -c -s "danbooru.donmai.us" -t "rating:safe" --return-images -i 20 -m 20

What do I expect:

Grabber outputs the URLs to the console, exits.

What happens:

The control returns immediately to the command line window, Grabber.exe runs in the background, invisible. If I add command-line redirection ``>temp.txt`` then eventually it prints out the results there, but never to the console.

Why it matters:

Order of operations is important for scripting, as is returned error code. Plus, ``--debug`` doesn't work this way.

**Comment by himselfv** [comment]
Oh, it seems this is not the first time I've stumbled upon this, and there was a workaround in my scripts. I hope this is not the second time I'm filing this, then.

Workaround: redirect output to ``tee`` from unix-utils:

> Grabber -c -s "danbooru.donmai.us" -t "rating:safe" --return-images -i 20 -m 20 | tee

This does not seem to redirect ``--debug`` though. Probably it goes to another pipe.

**Comment by Bionus** [comment]
> Grabber outputs the URLs to the console, exits.

That's exactly what's actually supposed to happen indeed (and what I get on Windows). I'll investigate on Linux, but seeing the source code, all the networking is indeed done in the background, I'll reproduce and fix this.

```
./Grabber.exe -c -s "danbooru.donmai.us" -t "rating:safe" --return-images -i
http://danbooru.donmai.us/data/0f52f01b52e63d16ae7d4f064e4bc31c.jpg
http://danbooru.donmai.us/data/27d9b8a9ce2ce99cb9231138048f0122.jpg
http://danbooru.donmai.us/data/f740760298169b7209cc8cfc5e2dac20.png
http://danbooru.donmai.us/data/9b6986e0d3ca8fc4cb87fb465a213108.png
http://danbooru.donmai.us/data/8d89bf6b3b50987f467c83df55b330d4.jpg
http://danbooru.donmai.us/data/95aecbf8a22e29e78c6e919de13d576e.jpg
http://danbooru.donmai.us/data/4a45c983b98bb02eb23425136db88dfa.png
http://danbooru.donmai.us/data/46bf5e509d9e102e99f4be3233844f00.jpg
http://danbooru.donmai.us/data/c4a95d452d5c8329d7f19351cec26868.jpg
http://danbooru.donmai.us/data/c303e62f0086ad056096d5a1584596f3.jpg
http://danbooru.donmai.us/data/79f0663ab44d7925b6301cc565cc027d.png
http://danbooru.donmai.us/data/34c8b94f328a96853ddae3797575fea4.jpg
http://danbooru.donmai.us/data/8d16a738b45b546612d1b8c536ca88c9.png
http://danbooru.donmai.us/data/ce91286935eea61317137e4446bb5bb7.jpg
http://danbooru.donmai.us/data/ade117dd23b6511bb10340aa42e8c787.jpg
http://danbooru.donmai.us/data/857e36bfbd2aca7e3c79c698857fbb82.jpg
http://danbooru.donmai.us/data/90e9617bdd19cf115071972940ff6628.png
http://danbooru.donmai.us/data/bcbca8cc0afe8fd1a1e1bcc1f2c4b139.jpg
http://danbooru.donmai.us/data/0f080fff2cff83f7f3db01f5b47ffeaa.jpg
```

> Plus, --debug doesn't work this way.

It's called --debug for some reason, but it should actually be called -v/--verbose.

**Comment by himselfv** [comment]
I'm on Windows too (Windows 10). Maybe we're different in that you're setting CONFIG += use_cli? It's not set in release builds so now that I think about it, that's probably why it behaves like that.

------

I've investigated a bit if there's a way to make a GUI app behave in a console-like fashion and written this wrapper:

http://pastebin.com/2KaCfXQK

It's a console app which is placed besides grabber.exe and runs it, passing all the parameters, translates its output and waits for it to return. It acts as a command line version almost in all respects.

Consider it a contribution if you need it, or take whatever you want out of it.

There's one thing that can be improved. By default, stdout/stderr in C Runtime are unbuffered. But when CRT detects that the output is piped, it turns on buffering, so the results don't appear in the pipe immediately.

Maybe you could call the following when running in the console mode? Or maybe make a separate flag for it?
```
setbuf(stdout, NULL);
setbuf(stderr, NULL);
```

It makes matters no worse when running straight from the console (with ``--use-cli``) because then CRT already does not use buffering. But it may make things suboptimal when writing straight to a file. So maybe there should be a distinction between what is now "--console" and this mode. Maybe not.

----

By the way, it's easier the other way around. If you build Grabber.exe as a console app first and foremost, then all it takes to launch it as a GUI app (without the flickering console window) is to launch it like this:

```
CreateProcess(appName, cmdLine, NULL, NULL, TRUE, DETACHED_PROCESS, NULL, NULL, &si, &pi)
```

Note the ``DETACHED_PROCESS``. This creates the process without the console window. In the Grabber, you can test it like this:

```
	if (GetConsoleWindow() == 0) {
		//GUI mode
	}
	else {
		//Console mode
	}
```

Here's the entire code for the GUI wrapper for this case -- most of it is rebuilding the command line:

http://pastebin.com/g1hfURLj

**Comment by Bionus** [comment]
> Maybe we're different in that you're setting CONFIG += use_cli?

I always use the same configuration as in release to prevent that kind of issue.

The problem with your suggestion is that Grabber is made using Qt, you can see that the way windows are managed is totally different: [main.cpp](https://github.com/Bionus/imgbrd-grabber/blob/master/source/main.cpp).

All output is buffered (until std::endl) and sent by the Downloader class. The program loops until it receive the signal that all operations are finished:
```cpp
QEventLoop loop;
QObject::connect(dwnldr, SIGNAL(quit()), &loop, SLOT(quit()));
loop.exec();
```

When the download is finished, the Downloader emits the quit() signal, which is catched by the event loop which then exits (return 0).

I believe that you must be using windows's tty `cmd.exe` or PowerShell, in which I can indeed reproduce your issue, even in the latest version with the event loop. I'm personally using another shell and tty (cygwin / bash / mintty), so that's why I never noticed it.
Now that I can reproduce I'll do my best to find a fix and upload a 4.3.1 update.

> By the way, it's easier the other way around. If you build Grabber.exe as a console app first and foremost

That's what has been done with the CLI update, that's why i added the "use_cli" flag: without it, the program should not even once go through any GUI (the original objective was to even remove all these gui files from the final executable but I didn't have the time to do it).

**Comment by Bionus** [comment]
This will be fixed in 4.3.1, which will include two binaries: `Grabber.exe` (with a gui) and `Grabber.com` (with cli).  I chose the COM extension because `.com` files have higher priority when run from CLI (see [Wikipedia](https://en.wikipedia.org/wiki/COM_file#Execution_preference).

![cli-4 3 1](https://cloud.githubusercontent.com/assets/882719/13724126/5e12164c-e87b-11e5-9f4c-f55f85101045.png)

Note that the -c will not be necessary, unless you use the `.exe` file.

**Comment by himselfv** [comment]
That's one way to solve it indeed. Thanks!
